functions/go/set-image/transformer/images_transformer.go

```diff
@@ -0,0 +1,199 @@
+	// ONLY for kustomize, AdditionalImageFields is the user supplied fieldspec
+	AdditionalImageFields types.FsSlice `json:"additionalImageFields,omitempty" yaml:"additionalImageFields,omitempty"`
+	// resultCount logs the total count image change
+	resultCount int
+}
+
+// Run implements the Runner interface that transforms the resource and log the results
+func (t SetImage) Run(ctx *fn.Context, functionConfig *fn.KubeObject, items fn.KubeObjects) {
+	err := t.configDefaultData()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+	err = t.validateInput()
+	if err != nil {
+		ctx.ResultErrAndDie(err.Error(), nil)
+	}
+
+	for _, o := range items {
+		var result []setImageResult
+		switch o.GetKind() {
+		case "Pod":
+			err, result = t.setPodContainers(o)
+		case "Deployment", "StatefulSet", "ReplicaSet", "DaemonSet", "PodTemplate":
+			err, result = t.setPodSpecContainers(o)
+		}
+		t.logResult(ctx, err, result, o)
+	}
+
+	if t.AdditionalImageFields != nil {
+		custom.SetAdditionalFieldSpec(functionConfig.GetMap("image"), items, functionConfig.GetSlice("additionalImageFields"), ctx)
+	}
+
+	summary := fmt.Sprintf("summary: updated a total of %v image(s)", t.resultCount)
+	ctx.ResultInfo(summary, nil)
+}
+
+// Config transforms the data from ConfigMap to SetImage struct
+func (t *SetImage) configDefaultData() error {
+	for key, val := range t.DataFromDefaultConfig {
+		switch key {
+		case "name":
+			t.Image.Name = val
+		case "newName":
+			t.Image.NewName = val
+		case "newTag":
+			t.Image.NewTag = val
+		case "digest":
+			t.Image.Digest = val
+		default:
+			return fmt.Errorf("ConfigMap has wrong field name %v", key)
+		}
+	}
+	return nil
+}
+
+// validateInput validates the inputs passed into via the functionConfig
+func (t *SetImage) validateInput() error {
+	// TODO: support container name and only one argument input in the next PR
+	if t.Image.Name == "" {
+		return fmt.Errorf("must specify `name`")
+	}
+	if t.Image.NewName == "" && t.Image.NewTag == "" && t.Image.Digest == "" {
+		return fmt.Errorf("must specify one of `newName`, `newTag`, or `digest`")
+	}
+	if t.Image.NewTag != "" && t.Image.Digest != "" {
+		return fmt.Errorf("image `newTag` and `digest` both set, set only one")
+	}
+	return nil
+}
+
+// updateContainerImages updates the images inside containers, return potential error, and a list of logging results
+func (t *SetImage) updateContainerImages(pod *fn.SubObject) (error, []setImageResult) {
+	var containers fn.SliceSubObjects
+	containers = append(containers, pod.GetSlice("iniContainers")...)
+	containers = append(containers, pod.GetSlice("containers")...)
+
+	var result []setImageResult
+	for _, o := range containers {
+		oldValue := o.NestedStringOrDie("image")
+		if !image.IsImageMatched(oldValue, t.Image.Name) {
+			continue
+		}
+		newName := getNewImageName(oldValue, t.Image)
+		if oldValue == newName {
+			return nil, nil
+		}
```

Comment:
I think the right logic should be using `continue` here rather than `return`.  And we should have a e2e test to cover this.